File: lib_common/src/main/java/com/yyxnb/common/LibraryInitializer.java
package com.yyxnb.common;

import android.content.ContentProvider;
import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ProcessLifecycleOwner;
import androidx.multidex.MultiDex;

import com.yyxnb.common.utils.log.LogUtils;

/**
 * 使用ContentProvider初始化三方库
 *
 * @author yyx
 */
public class LibraryInitializer extends ContentProvider {
    @Override
    public boolean onCreate() {
        // 初始化
        Context context = AppConfig.getInstance().getContext();

        if (context != null) {

            // 突破65535的限制
            MultiDex.install(context);

            // 应用监听
            ProcessLifecycleOwner.get().getLifecycle().addObserver(new AppLifeObserver());

            LogUtils.init()
                    //设置全局tag
                    .setTag("---What---")
                    //是否显示日志，默认true，发布时最好关闭
                    .setShowThreadInfo(AppConfig.getInstance().isDebug())
                    .setDebug(AppConfig.getInstance().isDebug());

        }


        return true;
    }

    @Nullable
    @Override
    public Cursor query(@NonNull Uri uri, @Nullable String[] projection, @Nullable String selection, @Nullable String[] selectionArgs, @Nullable String sortOrder) {
        return null;
    }

    @Nullable
    @Override
    public String getType(@NonNull Uri uri) {
        return null;
    }

    @Nullable
    @Override
    public Uri insert(@NonNull Uri uri, @Nullable ContentValues values) {
        return null;
    }

    @Override
    public int delete(@NonNull Uri uri, @Nullable String selection, @Nullable String[] selectionArgs) {
        return 0;
    }

    @Override
    public int update(@NonNull Uri uri, @Nullable ContentValues values, @Nullable String selection, @Nullable String[] selectionArgs) {
        return 0;
    }
}

File: arch/src/main/java/com/yyxnb/arch/LibraryInitializer.kt
package com.yyxnb.arch

import android.app.Application
import android.arch.lifecycle.ProcessLifecycleOwner
import android.content.ContentProvider
import android.content.ContentValues
import android.database.Cursor
import android.net.Uri
import android.support.multidex.MultiDex
import com.github.anzewei.parallaxbacklayout.ParallaxHelper
import com.yyxnb.arch.base.ManagerActivityLifecycleCallbacksImplI
import com.yyxnb.arch.utils.log.LogUtils
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import me.jessyan.autosize.AutoSizeConfig

/**
 * 使用ContentProvider初始化三方库
 */
class LibraryInitializer : ContentProvider() {

    /**
     * ContentProvider 的 onCreate() 是介于 Application 的 attachBaseContext(Context) 和 onCreate() 之间所调用的，
     * Application 的 attachBaseContext(Context) 方法被调用这就意味着 Application 的 Context 被初始化了，
     * 而 ContentProvider 拿到的 Context 也正就是 Application，所以可以在 ContentProvider 的 onCreate() 方法中完成相应的初始化操作 。
     */
    override fun onCreate(): Boolean {
        // 可以拿到ApplicationContext进行有些初始化操作
        GlobalScope.launch {
            (context?.applicationContext as Application).let {

                //突破65535的限制
                MultiDex.install(it)

                Arch.init(it)

                AutoSizeConfig.getInstance().isCustomFragment = true

                //系统会在每个 Activity 执行完对应的生命周期后都调用这个实现类中对应的方法
                it.registerActivityLifecycleCallbacks(ParallaxHelper.getInstance())
                it.registerActivityLifecycleCallbacks(ManagerActivityLifecycleCallbacksImplI())
                ProcessLifecycleOwner.get().lifecycle.addObserver(AppLifeObserver())

                LogUtils.init()
                        .setTag("Arch")//设置全局tag
                        .setShowThreadInfo(true).setDebug(Arch.isDebug) //是否显示日志，默认true，发布时最好关闭
            }
        }

        return true
    }

    override fun insert(uri: Uri, values: ContentValues?): Uri? {
        return null
    }

    override fun query(
            uri: Uri,
            projection: Array<String>?,
            selection: String?,
            selectionArgs: Array<String>?,
            sortOrder: String?
    ): Cursor? {
        return null
    }

    override fun update(uri: Uri, values: ContentValues?, selection: String?, selectionArgs: Array<String>?): Int {
        return 0
    }

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int {
        return 0
    }

    override fun getType(uri: Uri): String? {
        return null
    }

}